src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -244,7 +250,43 @@ private BoundExpression BindCompoundAssignment(AssignmentExpressionSyntax node,
             var leftConversion = CreateConversion(node.Left, leftPlaceholder, best.LeftConversion, isCast: false, conversionGroupOpt: null, best.Signature.LeftType, diagnostics);
 
             return new BoundCompoundAssignmentOperator(node, bestSignature, left, rightConverted,
-                leftPlaceholder, leftConversion, finalPlaceholder, finalConversion, resultKind, originalUserDefinedOperators, leftType, hasError);
+                leftPlaceholder, leftConversion, finalPlaceholder, finalConversion, resultKind, originalUserDefinedOperators, getResultType(left, forceDynamicResult), hasError);
+
+            TypeSymbol getResultType(BoundExpression left, bool forceDynamicResult)
```

Comment:
Would it be possible to use a common helper (similar to `AdjustAssignmentTarget`) for adjusting the return type of assignments? (maybe `AdjustAssignmentType`?)